stubs/python-decouple/decouple.pyi

```diff
@@ -0,0 +1,105 @@
+def strtobool(value: str | bool) -> bool: ...  # undocumented
+
+class UndefinedValueError(Exception): ...
+class Undefined: ...
+
+undefined: Undefined
+
+class Config:
+    repository: RepositoryEmpty
+    def __init__(self, repository: RepositoryEmpty) -> None: ...
+    @overload
+    def get(self, option: str, default: str = ...) -> str: ...
+    @overload
+    def get(self, option: str, cast: Callable[..., _T]) -> _T: ...
+    @overload
+    def get(self, option: str, default: str, cast: Callable[..., _T]) -> _T: ...
+    @overload
+    def __call__(self, option: str, default: str = ...) -> str: ...
+    @overload
+    def __call__(self, option: str, cast: Callable[..., _T]) -> _T: ...
+    @overload
+    def __call__(self, option: str, default: str, cast: Callable[..., _T]) -> _T: ...
+
+class RepositoryEmpty:  # undocumented
+    def __init__(self, source: str = ..., encoding: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class RepositoryIni(RepositoryEmpty):
+    SECTION: str
+    parser: ConfigParser
+    def __init__(self, source: str = ..., encoding: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class RepositoryEnv(RepositoryEmpty):
+    data: dict[str, str]
+    def __init__(self, source: str = ..., encoding: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class RepositorySecret(RepositoryEmpty):
+    data: dict[str, str]
+    def __init__(self, source: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class AutoConfig:
+    SUPPORTED: OrderedDict[str, RepositoryEmpty]
+    encoding: str
+    search_path: str | None
+    config: Config | None
+    def __init__(self, search_path: str | None = None) -> None: ...
+    @overload
+    def __call__(self, option: str, default: str = ...) -> str: ...
+    @overload
+    def __call__(self, option: str, cast: Callable[..., _T]) -> _T: ...
+    @overload
+    def __call__(self, option: str, default: str, cast: Callable[..., _T]) -> _T: ...
+
+config: AutoConfig
+
+class _CsvType(Protocol[_T, _TCsv]):
+    cast: Callable[..., _T]
+    delimiter: str
+    strip: str
+    post_process: Callable[..., _TCsv]
+    def __call__(self, value: str) -> _TCsv: ...
+
+@overload
+def Csv() -> _CsvType[str, list[str]]: ...
+@overload
+def Csv(cast: Callable[..., _T], delimter: str = ..., strip: str = ...) -> _CsvType[_T, list[_T]]: ...
+@overload
+def Csv(cast: Callable[..., _T], delimiter: str, strip: str, post_process: Callable[..., _TCsv]) -> _CsvType[_T, _TCsv]: ...
+@overload
+def Csv(
+    *, cast: Callable[..., _T], post_process: Callable[..., _TCsv], delimiter: str = ..., strip: str = ...
+) -> _CsvType[_T, _TCsv]: ...
```

Comment:
This is likely the most complex use case in this module. The Csv class exhibits default behavior that can be modified based on parameters passed into its `__init__` method. By default, `cast` is set to `str`, and `post_process` is set to `list`. Therefore, if these parameters are explicitly defined, the types provided should be used. However, if they are not specified, the return type of the `__call__` method should default to `list[str]`.

Reply:
If there is a better way to express this, I am open to suggestions.

Reply:
I haven't looked deeply, but it sounds like you simply need a generic `__init__` method.